mantis-control-plane/mantis-control-plane-server/src/main/java/io/mantisrx/master/events/WorkerMetricsCollector.java

```diff
@@ -0,0 +1,213 @@
+
+            switch (workerState) {
+                case Accepted:
+                    // do nothing; This is the initial state
+                    break;
+                case Launched:
+                    // this represents the scheduling time
+                    workerMetrics.reportSchedulingDuration(
+                        Math.max(0L, workerStatusEvent.getTimestamp() - metadata.getAcceptedAt()));
+                    break;
+                case StartInitiated:
+                    // do nothing; this event gets sent when the worker has received the request; it's too granular and expected to be really low -
+                    // so there's no point in measuring it.
+                    break;
+                case Started:
+                    workerMetrics.reportWorkerPreparationDuration(
+                        Math.max(0L, workerStatusEvent.getTimestamp() - metadata.getLaunchedAt()));
+                    break;
+                case Failed:
+                case Completed:
+                    workerMetrics.reportRunningDuration(
+                        Math.max(0L, workerStatusEvent.getTimestamp() - metadata.getStartedAt()));
+                    break;
+                case Noop:
+                    break;
+                case Unknown:
+                    log.error("Unknown WorkerStatusEvent {}", workerStatusEvent);
+                    break;
+            }
+        } catch (Exception e) {
+            log.error("Failed to process worker status event {}", workerStatusEvent);
+        }
+    }
+
+    private void cleanUp(JobId jobId) {
+        synchronized (jobsToBeCleaned) {
+            jobsToBeCleaned.add(new CleanupJobEvent(jobId, clock.instant()));
+        }
+    }
+
+    private class WorkerMetrics {
+
+        private final String clusterName;
+        private final Timer schedulingDuration;
+        private final Timer preparationDuration;
+        private final Timer runningDuration;
+
+        public WorkerMetrics(final String clusterName) {
+            this.clusterName = clusterName;
+            MetricGroupId metricGroupId =
+                new MetricGroupId("WorkerMetricsCollector", Tag.of("cluster", clusterName));
+            Metrics m = new Metrics.Builder()
+                .id(metricGroupId)
+                .addTimer(SCHEDULING_DURATION)
+                .addTimer(PREPARATION_DURATION)
+                .addTimer(RUNNING_DURATION)
+                .build();
+
+            m = MetricsRegistry.getInstance().registerAndGet(m);
+            this.schedulingDuration = m.getTimer(SCHEDULING_DURATION);
+            this.preparationDuration = m.getTimer(PREPARATION_DURATION);
+            this.runningDuration = m.getTimer(RUNNING_DURATION);
+        }
+
+        private void reportSchedulingDuration(long durationInMillis) {
+            this.schedulingDuration.record(durationInMillis, TimeUnit.MILLISECONDS);
```

Comment:
do we want to tag jobId etc.?

Reply:
I don't think atlas would be happy about that - atlas only supports a max cardinality of 1000. The goal is to look at metrics per resource cluster / mesos. For instance, we want to show that titus metrics are better or at least similar to mesos metrics.